### PR TITLE
Initial project pattern

### PR DESCRIPTION
**Pull request by valeriy-simakov**
https://github.com/networkservicemesh/cmd-forwarder-sriov/issues/3
Initial abstract forwarder application with dockerfile

### CODE REVIEW

File: src/forwarder/internal/pkg/cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	cmd := rootCmd
+	Flags(cmd.Flags())
+	cobra.OnInitialize(flags.FromEnv(flags.EnvPrefix, flags.EnvReplacer, cmd.Flags()))
+}
+
+var rootCmd = &cobra.Command{
+	Use:   "forwarder",
+	Short: "Provides xconnect network service",
+	Long: `Provides xconnect network service.  Supported mechanisms:
+     - memif
```
Comment (denis-tingajkin) [inline src/forwarder/internal/pkg/cmd/root.go:38]: Is this forwarder support these mechanisms?

File: src/forwarder/main_test.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+// limitations under the License.
+
+package main_test
+
+import (
+	"github.com/sirupsen/logrus"
+	"testing"
+)
+
+func TestCheck(t *testing.T) {
```
Comment (denis-tingajkin) [inline src/forwarder/main_test.go:24]: Why we need this test?

File: src/forwarder/internal/pkg/authz/write.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	"io/ioutil"
+
+	"github.com/open-policy-agent/opa/rego"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+)
+
+const (
+	// AuthzRegoFilename - default file containing authz rego policy
+	AuthzRegoFilename = "/etc/nsm/authz.rego"
```
Comment (denis-tingajkin) [inline src/forwarder/internal/pkg/authz/write.go:31]: Is this policy missed? Do not see authz.rego in these changes.

**Comment by denis-tingajkin** [comment]
Also please add signoff https://github.com/networkservicemesh/cmd-forwarder-sriov/pull/1/checks?check_run_id=712681479

**Comment by denis-tingajkin** [comment]
Also, I think we need to add CI for this repository. I've added an example of how you can add CI https://github.com/networkservicemesh/cmd-forwarder-sriov/issues/2#issuecomment-634579136

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,27 @@
+module github.com/networkservicemesh/cmd-forwarder-sriov
+
+go 1.14
+
+require (
+	github.com/OneOfOne/xxhash v1.2.8 // indirect
```
Comment (edwarnicke) [inline go.mod:6]: Please trim out the // indirect lines, they are usually an artifact of go get pollution.

**Comment by edwarnicke** [comment]
Please go have a look at:

https://github.com/networkservicemesh/cmd-forwarder-vppagent/pull/17

File: src/forwarder/internal/pkg/authz/write.go (hunk 3 above)
Comment (valeriy-simakov) [reply src/forwarder/internal/pkg/authz/write.go:31]: I did't see this file in 'cmd-forwarder-vppagent', so I thought that the user creates this file.

Comment (denis-tingajkin) [reply src/forwarder/internal/pkg/authz/write.go:31]: ok, then let's remove this const

File: src/forwarder/internal/pkg/authz/write.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2020 Cisco and/or its affiliates.
```
Comment (lvfxx) [inline src/forwarder/internal/pkg/authz/write.go:1]: Should we use `Copyright (c) 2020 Doc.ai and/or its affiliates`?

Comment (denis-tingajkin) [reply src/forwarder/internal/pkg/authz/write.go:1]: yes, we need to use this pattern

File: src/forwarder/internal/pkg/cmd/env.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2020 Cisco and/or its affiliates.
+//
```
Comment (haiodo) [inline src/forwarder/internal/pkg/cmd/env.go:2]: Ed already moved from corba to direct app for forwarder, so we should do same.

File: src/forwarder/internal/pkg/cmd/flags.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	"github.com/networkservicemesh/sdk/pkg/tools/flags"
+	"github.com/spf13/pflag"
+	"net/url"
+)
+
+// Name of the endpoint
+var Name string
+
+// ListenOnURL - where the forwarder should listen
+var ListenOnURL url.URL
```
Comment (haiodo) [inline src/forwarder/internal/pkg/cmd/flags.go:29]: Please check how Config is defined for new forwarder.

File: local/sdk-sriov/pkg/networkservice/chains/xconnectns/server.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+//             -name - name of the Forwarder
+//             -authzPolicy - policy for allowing or rejecting requests
+//             -tokenGenerator - token.GeneratorFunc - generates tokens for use in Path
+//             -clientUrl - *url.URL for the talking to the NSMgr
+//             -...clientDialOptions - dialOptions for dialing the NSMgr
+func NewServer(name string, authzPolicy *rego.PreparedEvalQuery, tokenGenerator token.GeneratorFunc, clientURL *url.URL, clientDialOptions ...grpc.DialOption) endpoint.Endpoint {
+	rv := xconnectNSServer{}
+	rv.Endpoint = endpoint.NewServer(
+		name,
+		authzPolicy,
```
Comment (edwarnicke) [inline local/sdk-sriov/pkg/networkservice/chains/xconnectns/server.go:48]: You will need to update to match changes in sdk for authzService

Comment (denis-tingajkin) [reply local/sdk-sriov/pkg/networkservice/chains/xconnectns/server.go:48]: +1

Also, I think we need to add config ci config(for example https://github.com/networkservicemesh/sdk/blob/master/.github/workflows/update-sdk-vppagent.yaml)  to updating this repo

File: .github/workflows/ci.yaml [hunk 9]
```diff
@@ -0,0 +1,145 @@
+---
```
Comment (denis-tingajkin) [inline .github/workflows/ci.yaml:1]: Please rebase your PR on the latest master. It should have `.github/workflows/ci.yaml`

File: local/sdk-sriov/pkg/networkservice/chains/xconnectns/server.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/endpoint"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/clienturl"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/connect"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/adapters"
+	"github.com/networkservicemesh/sdk/pkg/tools/addressof"
+	"github.com/networkservicemesh/sdk/pkg/tools/token"
+	"google.golang.org/grpc"
+)
+
+type xconnectNSServer struct {
```
Comment (haiodo) [inline local/sdk-sriov/pkg/networkservice/chains/xconnectns/server.go:35]: Should we name it like sriovServer?

Comment (valeriy-simakov) [reply local/sdk-sriov/pkg/networkservice/chains/xconnectns/server.go:35]: I thought that we can change it when we have a corresponding endpoint realization.

**Pull request merged**